tsl/src/nodes/vector_agg/function/float48_accum_single.c

```diff
@@ -0,0 +1,307 @@
+#endif
+					   ),
+	};
+
+	ArrayType *result = construct_array(transdatums,
+										3,
+										FLOAT8OID,
+										sizeof(float8),
+										FLOAT8PASSBYVAL,
+										TYPALIGN_DOUBLE);
+
+	*out_result = PointerGetDatum(result);
+	*out_isnull = false;
+}
+
+/*
+ * Youngs-Cramer update for rows after the first.
+ */
+static pg_attribute_always_inline void
+FUNCTION_NAME(update)(const uint64 *valid1, const uint64 *valid2, const CTYPE *values, int row,
+					  double *N, double *Sx
+#ifdef NEED_SXX
+					  ,
+					  double *Sxx
+#endif
+)
+{
+	const CTYPE newval = values[row];
+	if (!arrow_row_both_valid(valid1, valid2, row))
+	{
+		return;
+	}
+
+	/*
+	 * This code follows float8_accum(), see the comments there.
+	 */
+#ifdef NEED_SXX
+	Assert(*N > 0.0);
+	const double tmp = newval * (*N + 1.0) - (*Sx + newval);
+	*Sxx += tmp * tmp / (*N * (*N + 1.0));
+#endif
+
+	*N = *N + 1.0;
+	*Sx = *Sx + newval;
+}
+
+/*
+ * Combine two Youngs-Cramer states following the float8_combine() function.
+ */
+static pg_attribute_always_inline void
+FUNCTION_NAME(combine)(double *inout_N, double *inout_Sx,
+#ifdef NEED_SXX
+					   double *inout_Sxx,
+#endif
+					   double N2, double Sx2
+#ifdef NEED_SXX
+					   ,
+					   double Sxx2
+#endif
+)
+{
+	const double N1 = *inout_N;
+	const double Sx1 = *inout_Sx;
+#ifdef NEED_SXX
+	const double Sxx1 = *inout_Sxx;
+#endif
+
+	if (unlikely(N1 == 0))
+	{
+		*inout_N = N2;
+		*inout_Sx = Sx2;
+#ifdef NEED_SXX
+		*inout_Sxx = Sxx2;
+#endif
+		return;
+	}
+
+	if (unlikely(N2 == 0))
+	{
+		*inout_N = N1;
+		*inout_Sx = Sx1;
+#ifdef NEED_SXX
+		*inout_Sxx = Sxx1;
+#endif
+		return;
+	}
+
+	const double combinedN = N1 + N2;
+	const double combinedSx = Sx1 + Sx2;
+#ifdef NEED_SXX
+	const double tmp = Sx1 / N1 - Sx2 / N2;
+	const double combinedSxx = Sxx1 + Sxx2 + N1 * N2 * tmp * tmp / combinedN;
+#endif
+
+	*inout_N = combinedN;
+	*inout_Sx = combinedSx;
+#ifdef NEED_SXX
+	*inout_Sxx = combinedSxx;
+#endif
+}
+
+#ifdef NEED_SXX
+#define UPDATE(valid1, valid2, values, row, N, Sx, Sxx)                                            \
+	FUNCTION_NAME(update)(valid1, valid2, values, row, N, Sx, Sxx)
+#define COMBINE(inout_N, inout_Sx, inout_Sxx, N2, Sx2, Sxx2)                                       \
+	FUNCTION_NAME(combine)(inout_N, inout_Sx, inout_Sxx, N2, Sx2, Sxx2)
+#else
+#define UPDATE(valid1, valid2, values, row, N, Sx, Sxx)                                            \
+	FUNCTION_NAME(update)(valid1, valid2, values, row, N, Sx)
+#define COMBINE(inout_N, inout_Sx, inout_Sxx, N2, Sx2, Sxx2)                                       \
+	FUNCTION_NAME(combine)(inout_N, inout_Sx, N2, Sx2)
+#endif
+
+static pg_attribute_always_inline void
+FUNCTION_NAME(vector_impl)(void *agg_state, int n, const CTYPE *values, const uint64 *valid1,
+						   const uint64 *valid2, MemoryContext agg_extra_mctx)
+{
+	/*
+	 * Vector registers can be up to 512 bits wide.
+	 */
+#define UNROLL_SIZE ((int) (512 / 8 / sizeof(CTYPE)))
```

Comment:
This is just a guesstimate that is specific to a particular unrolled loop, no deep meaning. So I'd keep it local. Maybe we can make it more architecture-specific, for now I just wrote a "least-effort" version that is vectorized at least somehow.

Reply:
I am just wondering if there's a standard header file we can import that tells us AVX-512 is available and there some "define" related to the unroll size that says "512".

Reply:
Not sure if there's an easy cross-platform way to do it. Generally we want to unroll to the biggest available size across architectures, if there's only a smaller width available it will still lead to vectorizable code.